packages/ra-ui-materialui/src/input/AutocompleteInput.tsx

```diff
@@ -221,6 +221,16 @@ export const AutocompleteInput = (props: AutocompleteInputProps) => {
         [input.value, getSuggestionFromValue]
     );
 
+    const doesQueryMatchSuggestion = useMemo(() => {
+        if (typeof optionText !== 'string') {
+            return false;
+        }
```

Comment:
I think we should leverage `matchSuggestion`. Extract from the documentation:

> `matchSuggestion`: Required if `optionText` is a React element. Function returning a boolean indicating whether a choice matches the filter. `(filter, choice) => boolean`

Reply:
If `optionText` is not a React element but a function, we should call it and apply the same logic to its result (check if it is an element or a string and call `matchSuggestion` if needed?)

Reply:
Great, i've updated the PR. I've added videos for all scenarios, but please note that using `matchSuggestion` seemed to break the `useSuggestion` behaviour. Is this expected, or did I do something wrong?

I've left notes in the MR description

Reply:
One of the E2E tests failed, which flagged the issue I mentioned in my previous comment. I've resolved the issue, and the tests pass!